### PR TITLE
mds: recursively count opened files beneath a directory

### PR DESCRIPTION
**Pull request by xxhdx1985126**
Fixes: http://tracker.ceph.com/issues/40681
Signed-off-by: Xuehan Xu <xuxuehan@360.cn>

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

File: src/mds/OpenFileTable.cc [hunk 1]
```diff
@@ -58,47 +58,52 @@ void OpenFileTable::get_ref(CInode *in)
...
     }
 
+    pin->recursive_opened_files++;
     in = pin;
   } while (in);
 }
 
 void OpenFileTable::put_ref(CInode *in)
 {
+  bool only_ajust_recursive_opened_files = false;
```
Comment (ukernel) [inline src/mds/OpenFileTable.cc]: never set only_ajust_recursive_opened_files to true.  what is it used for

Comment (xxhdx1985126) [reply src/mds/OpenFileTable.cc]: @ukernel Hi, sorry that I uploaded the wrong code, now it's right. Please take a look, thanks:-)

**Comment by xxhdx1985126** [comment]
@ukernel Hi, I've corrected a few things, please take a look:-) Thanks:-)

**Comment by ukernel** [comment]
changes in OpenFileTable::{get_ref, put_ref} are not balanced, which is clearly wrong. besides, you don't handle rename (notify_link, notify_unlink}.

**Comment by xxhdx1985126** [comment]
@ukernel Hi, thanks for your review:-)

It seems that notify_link happens when linking/renaming a dir/file, either of which indicates that there are clients whose "want" cap for the corresponding inode isn't empty, which further means to mds that the file/dir is opened. And likewise the notify_unlink. So I think we should take notify_link and notify_unlink in account. Am I right?

And in my test, I made a deep subtree /a/b/c/d/e/f/g, "cd"ed into it, then umount the cephfs, on the mds side, I can see that after the umount, all "recursive_opened_file" are 0. And after analyzing the code, I still can't find why "changes in OpenFileTable::{get_ref, put_ref} are not balanced".

However, I found that to count the actually "opened" files, dirfrag shouldn't be taken into account. So, I ruled out the add_dirfrag/remove_dirfrag case. Please take a look:-) Thanks:-)

The following is my test result:
After "cd" into "/a/b/c/d/e/f/g", before umount:
```
[xuxuehan@xxh01v build]$ ./bin/ceph daemon mds.a dump tree /|grep opened
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
        "recursive_opened_files": 0,
        "recursive_opened_files": 1,
        "recursive_opened_files": 2,
        "recursive_opened_files": 3,
        "recursive_opened_files": 4,
        "recursive_opened_files": 5,
        "recursive_opened_files": 6,
        "recursive_opened_files": 7,
```
After umount:
```
        "recursive_opened_files": 0,
        "recursive_opened_files": 0,
        "recursive_opened_files": 0,
        "recursive_opened_files": 0,
        "recursive_opened_files": 0,
        "recursive_opened_files": 0,
        "recursive_opened_files": 0,
        "recursive_opened_files": 0,
```

**Comment by ukernel** [comment]
for example: 

cd /a/b/c/d
mv /a/b/c /

**Comment by xxhdx1985126** [comment]
@ukernel
Hi, I've modified the code to address the "notify_link/notify_unlink" issue:-)
Please take a look:-) Thanks:-)

**Comment by xxhdx1985126** [comment]
jenkins retest signed

**Comment by xxhdx1985126** [comment]
jenkins test signed

**Comment by xxhdx1985126** [comment]
jenkins test make check arm64

**Comment by xxhdx1985126** [comment]
please unmark

**Comment by xxhdx1985126** [comment]
please unmark

**Comment by gregsfortytwo** [comment]
@xxhdx1985126 sorry for the delay here, but this needs a rebase now. :(

@ukernel, this needs an updated review from you!

**Comment by ukernel** [comment]
I really don't understand why we need to track open files for whole directory hierarchy. In the tracker ticket, you said some clients may open massive files. you can use 'session ls' to check clients' caps counts (or number of caps that want Frw), and find these clients.

**Comment by xxhdx1985126** [comment]
@ukernel I think you are right, I'm closing this PR for now, thanks:-)

**Pull request closed without merging**